packages/grid/_modules_/grid/hooks/features/filter/useGridFilter.ts

```diff
@@ -172,7 +172,7 @@ export const useGridFilter = (
         const filterTreeNode = (
```

Comment:
The filtering and sorting logic for the tree data could be moved to a pre-processor to avoid shipping them as MIT.

In https://github.com/mui-org/material-ui-x/blob/2409df34aa924b4dd0d2ef80d59f2812974a1171/packages/grid/_modules_/grid/hooks/features/filter/useGridFilter.ts#L231

Would be smarter to exit the loop as soon a `node.depth > 0` is found? The tree nodes are ordered from the lowest depth to the highest.

Reply:
I think we should remain vigilant with pre-processor. We can have the temptation of answering any custom code with them, but they can make the global understanding of the codebase very complex.

For these two specific algorithm, I think it is better to start by in-lining them in `useGridFilter` and `useGridSorting` and then improve it (with pre-processor or another solution) when we want to use them in the row grouping feature.


> The tree nodes are ordered from the lowest depth to the highest.

Not necessarily.
Your auto generated nodes can be added after nodes of higher depth.

Reply:
> I think we should remain vigilant with pre-processor. We can have the temptation of answering any custom code with them, but they can make the global understanding of the codebase very complex.

I have a different vision. I found easier to understand when keeping everything related to a feature inside the same hook. The filtering hook is already quite long with 440 LoC in total, where 65 LoC are for the tree-data.

> For these two specific algorithm, I think it is better to start by in-lining them in useGridFilter and useGridSorting and then improve it (with pre-processor or another solution) when we want to use them in the row grouping feature.

Is it not just to copy and paste into a pre-processor? I think if we move to a pre-processor there's no other way then the pre-processor completely overriding the filtering logic.

Reply:
> I found easier to understand when keeping everything related to a feature inside the same hook

We agree on that point
My problem is more about when the behavior is triggered and how many time.

> where 65 LoC are for the tree-data.

The Tree Data or the Row Grouping, or the Aggregation (partially).
I did not put it in the Tree Data hook because it can be applied to any feature using a tree.
Same reason why it is `useGridRows` which store the tree structure.
My PR of the row grouping is very simple for the sorting / filtering / rows management because all the logic is compatible.

And when we will launch the row grouping for instance, the pre-processing change will potentially trigger several times if we change a shared prop (`props.groupingColDef` for instance) even if one of them is actually not enabled (`props.treeData = false`). And in that scenario, we will have to make sure that the filtering / sorting does not run several times.

A solution could be to introduce a `useTreeRows` which registers pre-processors only once is one of the tree features is enabled. I don't have a clear vision on that part yet.

Reply:
I think the pre-processing is a powerfull tool, but I also think we need more distance.
Things like "re-run all the pre-processors of a group whenever one processor said he has changed" is probably not suitable for extensive use.